File: solutions/110.balanced-binary-tree/balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
// top down
// const isBalanced = root => {
//   let flag = true

//   const walk = node => {
//     if (!node) return 0

//     let left = walk(node.left) + 1
//     let right = walk(node.right) + 1

//     if (Math.abs(left - right) > 1) flag = false

//     return Math.max(left, right)
//   }

//   walk(root)

//   return flag
// }


const isBalanced = root => {
  const walk = node => {
    if (!node) return 0

    let left = walk(node.left) + 1
    let right = walk(node.right) + 1

    if (left === 0 || right === 0) return -1

    if (Math.abs(left - right) > 1) return -1

    return Math.max(left, right)
  }

  return walk(root) !== -1
}

File: solutions/110.balanced-binary-tree/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        abs(left_depth, right_depth) > 1
        """

        def traverse(root):
            if not root:
                return 0

            left_height = traverse(root.left)
            right_height = traverse(root.right)

            if left_height is -1 or right_height is -1 or abs(left_height - right_height) > 1:
                return -1

            return max(left_height, right_height) + 1

        return traverse(root) is not -1
